### PR TITLE
Adds libversion.sh

### PR DESCRIPTION
**Pull request by rafariossaa**
This file contains a functions to handle version strings.

### CODE REVIEW

File: stretch/rootfs/libversion.sh [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    regex='([0-9]+)(\.([0-9]+)(\.([0-9]+))?)?'
+
+    if [[ $1 =~ $regex ]]; then
+        i=1
+        j=1
+        n=${#BASH_REMATCH[*]}
+
+        while [[ $i -lt $n ]]
+        do
+            if [ ! -z "${BASH_REMATCH[$i]}" ] && [ "${BASH_REMATCH[$i]:0:1}" != '.' ]
```
Comment (carrodher) [inline stretch/rootfs/libversion.sh]:
```suggestion
            if [ -n "${BASH_REMATCH[$i]}" ] && [ "${BASH_REMATCH[$i]:0:1}" != '.' ]
```
You have negated [`test -z`](https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html#Bash-Conditional-Expressions), resulting in a needless double-negative

File: stretch/rootfs/libversion.sh [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+            then
+                version[$j]=${BASH_REMATCH[$i]}
+                let j++
+            fi
+            let i++
+        done
+        if [ ! -z "$2" ]
```
Comment (carrodher) [inline stretch/rootfs/libversion.sh]:
```suggestion
        if [ -n "$2" ]
```
You have negated [`test -z`](https://www.gnu.org/software/bash/manual/html_node/Bash-Conditional-Expressions.html#Bash-Conditional-Expressions), resulting in a needless double-negative

File: stretch/rootfs/libversion.sh [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        i=1
+        j=1
+        n=${#BASH_REMATCH[*]}
+
+        while [[ $i -lt $n ]]
+        do
+            if [ ! -z "${BASH_REMATCH[$i]}" ] && [ "${BASH_REMATCH[$i]:0:1}" != '.' ]
+            then
+                version[$j]=${BASH_REMATCH[$i]}
+                let j++
```
Comment (carrodher) [inline stretch/rootfs/libversion.sh]:
```suggestion
                ((j++))
```
The `(( .. ))` arithmetic compound command evaluates expressions in the same way as to let, except it's not subject to glob expansion and therefore requires no additional quoting or escaping

File: stretch/rootfs/libversion.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+            fi
+            let i++
```
Comment (carrodher) [inline stretch/rootfs/libversion.sh]:
```suggestion
            ((i++))
```
The `(( .. ))` arithmetic compound command evaluates expressions in the same way as to let, except it's not subject to glob expansion and therefore requires no additional quoting or escaping

File: stretch/rootfs/libversion.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
+#
+# Library for managing versions strings
+
+# Functions
+########################
+# Gets version
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]:
```diff
- # Gets version
+ # Gets semantic version
```

File: stretch/rootfs/libversion.sh [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+########################
+# Gets version 
+# Arguments:
+#   $1 - string to extract major.minor.patch
+#   $2 - 1 to extract major, 2 to extract minor, 3 to extract patch
+# Returns:
+#   array with the major, minor and release
+#########################
+get_sematic_version () {
+    regex='([0-9]+)(\.([0-9]+)(\.([0-9]+))?)?'
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]: Add local variables to handle "required" parameters:

```diff
get_sematic_version () {
+  local version="${1:?version is required}"
```

Then adapt the function to this change. E.g.:

```diff
-    if [[ $1 =~ $regex ]]; then
+    if [[ "$version" =~ $regex ]]; then
```

You're not re-defining `regex` variable, the use "local readonly" variable

```diff
- regex='([0-9]+)(\.([0-9]+)(\.([0-9]+))?)?'
+ local -r regex='([0-9]+)(\.([0-9]+)(\.([0-9]+))?)?'
```

File: stretch/rootfs/libversion.sh [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        do
+            if [ -n "${BASH_REMATCH[$i]}" ] && [ "${BASH_REMATCH[$i]:0:1}" != '.' ]
+            then
+                version[$j]=${BASH_REMATCH[$i]}
+                ((j++))
+            fi
+            ((i++))
+        done
+        if [ -n "$2" ]
+        then
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]: Yoy're mixign the format:

```bash
if [[ ... ]]; then
```

With the format:

```bash
if [[ ... ]] 
then
```

Use the 1st one

File: stretch/rootfs/libversion.sh [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    regex='([0-9]+)(\.([0-9]+)(\.([0-9]+))?)?'
+
+    if [[ $1 =~ $regex ]]; then
+        i=1
+        j=1
+        n=${#BASH_REMATCH[*]}
+
+        while [[ $i -lt $n ]]
+        do
+            if [ -n "${BASH_REMATCH[$i]}" ] && [ "${BASH_REMATCH[$i]:0:1}" != '.' ]
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]: Use `[[ ... ]]` instead of `[ ... ]`

File: stretch/rootfs/libversion.sh [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+
+        while [[ $i -lt $n ]]; do
+            if [[ -n "${BASH_REMATCH[$i]}" ]] && [[ "${BASH_REMATCH[$i]:0:1}" != '.' ]];  then
+                version_sections[$j]=${BASH_REMATCH[$i]}
+                ((j++))
+            fi
+            ((i++))
+        done
+
+        if [ -n "$2" ]; then
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]: Improvement:
```diff
# Library for managing versions strings

+ # Load Generic Libraries
+ . /liblog.sh

...
-        if [ -n "$2" ]; then
+        number_regex='^[0-9]+$'
+        if [[ $2 =~ $number_regex ]] ; then
+            echo "${version_sections[$section]}"
+        else
+            stderr_print "section is not a number. Allowed values are: '1', '2', and '3'"
+            return 1
+        fi
```

File: stretch/rootfs/libversion.sh [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+get_sematic_version () {
+    local version="${1:?version is required}"
+    local section="${2:?section is required}"
+
+    local -r regex='([0-9]+)(\.([0-9]+)(\.([0-9]+))?)?'
+
+    if [[ "$version" =~ $regex ]]; then
+        i=1
+        j=1
+        n=${#BASH_REMATCH[*]}
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]: What is "BASH_REMATCH"?

File: stretch/rootfs/libversion.sh [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+
+        while [[ $i -lt $n ]]; do
+            if [[ -n "${BASH_REMATCH[$i]}" ]] && [[ "${BASH_REMATCH[$i]:0:1}" != '.' ]];  then
+                version_sections[$j]=${BASH_REMATCH[$i]}
```
Comment (juan131) [inline stretch/rootfs/libversion.sh]: I'd declare the variable `version_sections` at the top of the function as an empty array.

File: stretch/rootfs/libversion.sh (hunk 10 above)
Comment (rafariossaa) [reply stretch/rootfs/libversion.sh]: Is an array that is fetched when you do a regex test, like the one that is if.

http://molk.ch/tips/gnu/bash/rematch.html

**Pull request merged**